lib/src/command_runner/completion_command_runner.dart

```diff
@@ -65,26 +65,30 @@ abstract class CompletionCommandRunner<T> extends CommandRunner<T> {
     return _completionInstallation = completionInstallation;
   }
 
+  /// The list of commands that should not trigger the auto installation.
+  static const _reservedCommands = [
+    HandleCompletionRequestCommand.commandName,
+    InstallCompletionFilesCommand.commandName,
+    UnistallCompletionFilesCommand.commandName,
+  ];
+
   @override
   @mustCallSuper
   Future<T?> runCommand(ArgResults topLevelResults) async {
-    final reservedCommands = [
-      HandleCompletionRequestCommand.commandName,
-      InstallCompletionFilesCommand.commandName,
-      UnistallCompletionFilesCommand.commandName,
-    ];
-
-    final completionConfiguration = CompletionConfiguration.fromFile(
-      completionInstallation.completionConfigurationFile,
-    );
-    final isUninstalled = systemShell != null &&
-        completionConfiguration.uninstalls
-            .contains(command: executableName, systemShell: systemShell!);
-    if (enableAutoInstall &&
-        !reservedCommands.contains(topLevelResults.command?.name) &&
-        !isUninstalled) {
-      // When auto installing, use error level to display messages.
-      tryInstallCompletionFiles(Level.error);
+    if (enableAutoInstall && systemShell != null) {
+      final completionConfiguration = CompletionConfiguration.fromFile(
+        completionInstallation.completionConfigurationFile,
+      );
+      final isInstalled = completionConfiguration.installs
+          .contains(command: executableName, systemShell: systemShell!);
+      final isUninstalled = completionConfiguration.uninstalls
+          .contains(command: executableName, systemShell: systemShell!);
+      if (!_reservedCommands.contains(topLevelResults.command?.name) &&
+          !isUninstalled &&
+          !isInstalled) {
+        // When auto installing, use error level to display messages.
+        tryInstallCompletionFiles(Level.error);
+      }
```

Comment:
This logic is only related to the auto installation process which the CommandRunner is the only class aware of it.

Reply:
This library source has lots of logic in 3 main domains: 1. The command runner binding is the front-facing API to hook stuff from this library with the `args/command_runenr.dart` library.;2. The parser generates completion suggestions from user input; 3. All the installation logic involves lots of file reading and writing. 

It gets tough to wrap the head around the library's source code if those things are not separated. 


What if we send auto-install as a flag to the `tryInstallCompletionFiles` (which defaults to false)  so this can be abstracted from here to `CompletionIntallation` and `tryInstallCompletionFiles` can call 

`completionInstallation.canInstall()`

This is better because `CompletionInstallation` already has access to the following:
 - the system shell
 - `executableName`
 - `CompletionConfiguration` that should not be instantiated here (which also causes `completionInstallation.completionConfigurationFile` to be a public getter)

Reply:
I agree on the flag, we both added a comment stating the flag simultaneously, see https://github.com/VeryGoodOpenSource/cli_completion/pull/73#discussion_r1202715744.